File: 115. Distinct Subsequences.c
//ref: https://discuss.leetcode.com/topic/46346/my-0-ms-dp-solution-in-c-o-mn-time-o-n-space
#define MAXN 1024
int numDistinct(char* s, char* t) {
	const int lens = strlen(s), lent = strlen(t);
	int dp[MAXN] = {1};
	int i, j;

	for (i = 0; i < lens; i++)
		for (j = lent; j >= 0; j--)
			if (t[j] == s[i])
				dp[j + 1] += dp[j];
	return dp[lent];
}

File: 115. Distinct Subsequences/115. Distinct Subsequences.py
#https://discuss.leetcode.com/topic/6465/a-dp-solution-with-clarification-and-explanation/9
class Solution(object):
    def numDistinct(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: int
        """
        lent = len(t)
        dp = [1] + [0] * lent
        for i in xrange(len(s)):
            for j in reversed(xrange(lent)):
                if t[j] == s[i]:
                    dp[j + 1] += dp[j]
        return dp[-1]
